### PR TITLE
splits denoise into four methods

### PR DESCRIPTION
**Pull request by gregcaporaso**
The motivation for this is described in #14. 

It may or may not be worth merging this PR, as there are a couple of issues here that are outside the scope of what I'm able to help out with right now. (These issues weren't clear to me when I started working on this.)

First, it looks like the resulting feature table from these methods should really be the one that is currently called `final.only-16s.biom`. That file is not referenced at all in this plugin, so I think the filtering methods that I adapted here are returning the unfiltered feature table. 

Also, this PR doesn't include the negate option, but it sounds like that is necessary for these parameters to be used. Since it sounds like the usage of that parameter is not very intuitive and @amnona says they're planning to update it (#15), I think this parameter should not be exposed to users as a parameter in `q2-deblur`, but rather passed to `deblur` when necessary internally from these methods. This will avoid users misusing the parameter and not getting the results they expect. 

This PR does address the issues presented in #14 and #18, so if this is closed without merge those should be addressed in a new PR.

### CODE REVIEW

**Comment by wasade** [comment]
I'm not sure that this decomposition is what we want to do. For the time being, we anticipate that most uses of deblur will be for 16S, in which case it is not clear why a user would want to change the positive reference database. Additionally, since the positive reference database is now an input, I do not think it is possible to have it use a default value. What I propose instead is are two methods "denoise-16S" and "denoise-other" where the former does not allow for changing the positive reference and the latter allows for specifying one, and implicitly solves the default value issue.

From discussion with @amnona, we also think that the negative filter options should just be omitted. Having them introduces confusion, and it is unlikely that users will want to change that even when operating on other loci.

We are making changes to Deblur proper right now on the documentation, and revising a few of the exposed commands. That will trigger a version change, which will need to be reflected on this repo. The changes to the command structure can still happen in this PR -- I'll issue a PR against this PR shortly for them.

**Comment by gregcaporaso** [comment]
That all makes sense to me. Feel free to close this PR if you don't need it.

**Comment by wasade** [comment]
Easier to work off the work you've already put in. Want to give me push access to your repo and I can push into this branch? Changes almost there to satisfy the last comment

**Comment by wasade** [comment]
This PR has morphed a bit and now includes #17 as well.

**Comment by wasade** [comment]
@mortonjt @amnona @tanaes @RNAer @antgonza @josenavas possible for review? Jon/Zech, I added both of you as both of you have used the previous iteration of the plugin. Jamie/Antonio/Jose/Amnon, you guys have all been reviewing on the main project so comments on this wrapper would be appreciated. This should be in sync with deblur 0.1.9-dev as of ec8b1a7 -- we haven't cut the next release of the main project yet but that should be happening relatively soon.

@gregcaporaso I know you do not have time right now for a full review, but just to summarize and so you're aware, I brought this up to speed with the approach taken in the current iteration of q2-dada2 including adaption of its parameter validation and reused parts of its unit tests. In addition, there has been a large rework of the wrapped project's documentation including both its readme and the help text associated with the parameters. 

One important note: I opted to _not_ expose all of the files produced by Deblur. In my opinion, doing so would increase confusion without gain **for the majority of users** as the other files are not items generally feed into downstream components. The primary usecase that I foresee is to utilize the table composed of what Deblur determines to be high quality, where quality is based off of the application of the different thresholds, filtering for artifacts, and coarse recruitment to a loci specific reference. One possible precedent here is that q2-dada2 does not produce a "failures" artifact, although an immediate counterpoint is that q2-dada2 does expose a quality visualization method. The decision about what outputs are exposed can of course be revised, or we can proceed forward as is and assess whether other outputs should be generated based on user feedback.

File: q2_deblur/plugin_setup.py [hunk 1]
```diff
@@ -97,52 +28,89 @@ def denoise(demultiplexed_seqs: SingleLanePerSampleSingleEndFastqDirFmt,
...
-        'min_size': ("In each sample, discard all features with an abundance "
-                     "less than min_size."),
-        'negate': ("Discard (rather than retain) all sequences aligning to "
-                   "the sequences provided in neg_ref_fp."),
-        'jobs_to_start': "Number of jobs to start (if to run in parallel).",
-        'hashed_feature_ids': "If true, hash the feature IDs."
+    parameter_descriptions=_parameter_descriptions,
+    output_descriptions=_output_descriptions,
+    name='Deblur sequences.',
+    description=('Perform sequence quality control using the deblur workflow '
```
Comment (antgonza) [inline q2_deblur/plugin_setup.py]: perhaps worth adding and making super clear that this doesn't affect the results for other target genes?

File: q2_deblur/plugin_setup.py [hunk 2]
```diff
@@ -97,52 +28,89 @@ def denoise(demultiplexed_seqs: SingleLanePerSampleSingleEndFastqDirFmt,
...
+                           "sequences aligning to these sequences."),
     },
-    name='Perform sequence quality control using the deblur workflow',
-    description='Apply the deblur quality control workflow'
+    parameter_descriptions=_parameter_descriptions,
+    output_descriptions=_output_descriptions,
+    name='Deblur sequences and retain those similar to the reference.',
+    description=('Perform sequence quality control using the deblur workflow, '
+                 'including positive alignment-based filtering. This mode of '
+                 'execution is particularly useful when operating on non-16S '
```
Comment (antgonza) [inline q2_deblur/plugin_setup.py]: Perhaps similar to this?

**Comment by wasade** [comment]
It isn't clear to me at this time for how to handle reverse reads. For Deblur proper, the approach is to join the reads prior to execution, so by extension, that would place it upstream of this plugin.

File: q2_deblur/_denoise.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+_NAT_NUM = (lambda x: x > 0, 'greater than zero')
+_SKIP = (lambda x: True, '')
+_valid_inputs = {
+    'trim_length': _GTE_NEG_1,
+    'mean_error': _NAT_NUM,
+    'indel_prob': _NAT_NUM,
+    'indel_max': _WHOLE_NUM,
+    'min_reads': _WHOLE_NUM,
+    'min_size': _WHOLE_NUM,
+    'jobs_to_start': _WHOLE_NUM,
```
Comment (josenavas) [inline q2_deblur/_denoise.py]: Shouldn't it be `_NAT_NUM`?

**Comment by antgonza** [comment]
perhaps the question is, what will happen if we encounter a MANIFEST with reverse reads? I guess they should be ignored or raise an error/warning ...

**Comment by gregcaporaso** [comment]
> perhaps the question is, what will happen if we encounter a MANIFEST with reverse reads? I guess they should be ignored or raise an error/warning ...

That's [what happens](https://github.com/qiime2/q2-dada2/blob/5ca0103b9fa8ae33ec9ab9829af266fffabe8fdb/q2_dada2/plugin_setup.py#L36) in q2-dada2 if you call `denoise-single` with paired-end reads, so you could model this from that. (If you want to denoise paired-end reads, you use `denoised-paired`, but we didn't want users to have to create both a `SampleData[SequencesWithQuality]` and a `SampleData[PairedEndSequencesWithQuality]` if they wanted to (for example) compare them.

File: q2_deblur/plugin_setup.py [hunk 4]
```diff
@@ -97,52 +28,89 @@ def denoise(demultiplexed_seqs: SingleLanePerSampleSingleEndFastqDirFmt,
...
+                 '88% OTUs from Greengenes 13_8. The reference is used to '
+                 'assess whether each sequence is likely to be 16S by a local '
+                 'alignment using SortMeRNA with a permissive e-value.')
+)
+
+
+plugin.methods.register_function(
+    function=q2_deblur.denoise_other,
+    inputs={
+        'demultiplexed_seqs': SampleData[SequencesWithQuality],
```
Comment (josenavas) [inline q2_deblur/plugin_setup.py]: Is it forced to have the Quality information for deblur to run? Or am I misinterpreting the meaning of SequencesWithQuality?

**Comment by wasade** [comment]
@gregcaporaso oh nice, so q2 will gracefully resolve to just the fwd reads?

File: q2_deblur/_denoise.py (hunk 3 above)
Comment (wasade) [inline q2_deblur/_denoise.py]: Yes, good catch

File: q2_deblur/plugin_setup.py (hunk 4 above)
Comment (wasade) [inline q2_deblur/plugin_setup.py]: Deblur does not use the quality data, however, Deblur is also only defined over MiSeq and HiSeq, so quality scores are absolutely feasible here. Unclear to me if we should just require it, as the lack of quality scores increases the potential for the method to be used on non-illumina data.

**Comment by gregcaporaso** [comment]
> @gregcaporaso oh nice, so q2 will gracefully resolve to just the fwd reads?

Yes, exactly. 

Also, note that `SampleData[SequencesWithQuality]` doesn't imply single-end reads - it could also contain reads that have already been joined. So, if you end up having a paired-end joiner run before `deblur denoise`, the results could be placed in a `SampleData[SequencesWithQuality]` that could then be passed to this method. (That's how we're thinking about this now at least, but we also don't have a stand-alone paired-end joiner at the moment, so we're open to other ideas here if something else makes more sense.)

**Comment by wasade** [comment]
Okay, excellent, thank you

File: q2_deblur/plugin_setup.py (hunk 1 above)
Comment (wasade) [inline q2_deblur/plugin_setup.py]: I revised the text a bit more, I hope that helps?

**Pull request merged**